lib/test_runner.zig

```diff
@@ -61,23 +61,24 @@ pub fn main() void {
         if (!have_tty) {
             std.debug.print("{d}/{d} {s}... ", .{ i + 1, test_fn_list.len, test_fn.name });
         }
-        const result = if (test_fn.async_frame_size) |size| switch (io_mode) {
-            .evented => blk: {
-                if (async_frame_buffer.len < size) {
-                    std.heap.page_allocator.free(async_frame_buffer);
-                    async_frame_buffer = std.heap.page_allocator.alignedAlloc(u8, std.Target.stack_align, size) catch @panic("out of memory");
-                }
-                const casted_fn = @ptrCast(fn () callconv(.Async) anyerror!void, test_fn.func);
-                break :blk await @asyncCall(async_frame_buffer, {}, casted_fn, .{});
-            },
-            .blocking => {
-                skip_count += 1;
-                test_node.end();
-                progress.log("SKIP (async test)\n", .{});
-                continue;
-            },
-        } else test_fn.func();
-        if (result) |_| {
+        if (result: {
```

Comment:
If this change is necessary then that is one more downside for this more complex approach.

Reply:
The idea is that without this restriction, internally to a function error return traces can be left un-popped.

Here's what it would look like if we relaxed it:
```zig
const x = foo(); // fails
const y = baz(); // also fails
switch (cond) {
    0 => x,
    1 => y,
} catch {
    if (cond) {
        return error.Foo;  // both foo() and baz() end up in the error return trace, followed by this line
    }
    // A non-error return will still pop both foo() and baz(), thankfully
    return;            
} 

try quaf();                       // If this fails, error trace will show foo(), followed by baz(), followed by quaf()
if (cond2)
    return error.UnrelatedError;  // both foo() and baz() end up in the error return trace, followed by this line

// A non-error return can will still pop both foo() and baz(), thankfully
return;
```

See the commit message of 09842f9 for more info.

One other thought I had was to add a bit to errors so that we can notice when one's trace was killed due to being saved to a variable and to augment to panic message with "(error return trace includes dropped entries -- see `<link>` for more info)"

Reply:
I'm not sure what the downside in that example is supposed to be?

Reply:
The problem I see is that `foo()` and `baz()` pollute the error return trace even if they are "handled".

For example, in a loop:
```zig
while (i < N) : (i += 1) {
    const x = foo();
    if (x) |_| {
        // code here
    } else |err| {
        if (cond)
            return error.Foo; // Return trace includes foo() up to `i` times in the error trace
    }
}
``` 
`foo()` will appear in the error return trace however many times it failed during this function call.

This may be a matter of opinion, though. I'm happy to modify the behavior if desired.

Reply:
There is a different relaxation that might make sense: when initializing `const`/`var`, delay popping an error trace for the initializer until the variable falls out of scope. This wouldn't resolve my first example, but it would fix the second.

Using the same technique for args (delay pop until function returns a non-error) should resolve the problematic example below too.

I'll experiment a bit and report back

Reply:
Sorry for the delay here, just got back from a trip a few days ago 🏖️ 

I've updated this PR to include the fix for function args, and `testing.expectError` now works as expected.

Keeping error traces alive for a `const` variable (until the end of the containing block) still seems like a good idea. However, it requires moving significant portions of this logic to Sema, so I may wait for some feedback before I dive in.